[CodeCompletion] Avoid type checking all top-level decls in the primary file

rintaro opened the pull request:
The first commit is purely bug fixes. They are revealed if the second commit is enabled. However they can be reproduced if the declaration is in the other files in the current module. (Added `test/IDE/complete_multifile.swift` regression test case).

- Default memberwise initializer for struct was not suggested
- Dynamic member lookup didn't work
- `init(rawValue:)` was not suggested for `RawRepresentable` types (rdar://problem/56391233)
- '$name' was not suggested for `@propertyWrapper`ed value

The second commit actually stops eager typechecking decls in the primary file.

- Use `performParseAndResolveImportsOnly()` to invoke the frontend
- Do `bindExtensions()` in `ide::typeCheckContextUntil()`
- Typecheck preceding `TopLevelCodeDecl`s only if the compleiton is in
  a `TopLevelCodeDecl`
- Other related tweaks

rdar://problem/56636747

rintaro committed:
[CodeCompletion] Fix several completion failure in multi-file scenario

- Default memberwise initializer for struct was not suggested
- Dynamic member lookup didn't work
- `init(rawValue:)` was not suggested for `RawRepresentable` types
- '$name' was not suggested for `@propertyWrapper`ed value

rdar://problem/56391233

rintaro committed:
[CodeCompletion] Avoid typechecking all toplevel decls in the current file

- Use `performParseAndResolveImportsOnly()` to invoke the frontend
- Do `bindExtensions()` in `ide::typeCheckContextUntil()`
- Typecheck preceding `TopLevelCodeDecl`s only if the compleiton is in
  a `TopLevelCodeDecl`
- Other related tweaks

rdar://problem/56636747

rintaro commented on lib/IDE/CodeCompletion.cpp:
```diff
@@ -4092,10 +4092,46 @@ class CompletionOverrideLookup : public swift::VisibleDeclConsumer {
+            Access = std::max(Access, Requirement->getFormalAccess());
+        }
+      }
+    }
```
Previously `B.foo()` is `overwriteAccess()` from `diagnoseWitnessFixAccessLevel()`. Since the typechecking is now in `DiagnosticSuppression`, `diagnoseWitnessFixAccessLevel()` is not called so we need to manually check it.

rintaro commented on lib/IDE/CodeCompletion.cpp:
```diff
@@ -4318,9 +4354,11 @@ class CompletionOverrideLookup : public swift::VisibleDeclConsumer {
+      return;
+
+    // A 'class' member with an initial value cannot be overriden either.
+    if (D->isStatic() && isa<VarDecl>(D) && cast<VarDecl>(D)->hasInitialValue())
```
`HasInitialValueAttr` is injected in the typechecking.

rintaro commented on lib/IDE/CodeCompletion.cpp:
```diff
@@ -5096,6 +5135,9 @@ void CodeCompletionCallbacksImpl::doneParsing() {
       CurDeclContext->getASTContext().SourceMgr.getCodeCompletionLoc());
 
+  // Add keywords even if type checking fails completely.
+  addKeywords(CompletionContext.getResultSink(), MaybeFuncBody);
```
Moved after `typeCheckContextUntil()` because `addKeywords()` may call `getSelfProtocolDecl()` which can only be called after `bindExtensions()`.

rintaro commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -1145,6 +1145,8 @@ visitDynamicCallableAttr(DynamicCallableAttr *attr) {
                                       Identifier expectedLabel,
                                       bool ignoreLabel = false) {
+  // FIXME(InterfaceTypeRequest): Remove this.
+  (void)decl->getInterfaceType();
```
`index->hasInterfaceType()` few lines below needs this.

CodaFi replied:
After #27922 you can delete this.

rintaro replied:
Indeed. Thanks!

rintaro commented:
@swift-ci Please smoke test

rintaro commented:
@swift-ci Please smoke test

rintaro commented:
@swift-ci Please test Linux

rintaro commented:
@swift-ci Please test Linux

rintaro commented:
@swift-ci Please smoke test

benlangmuir commented on lib/Sema/LookupVisibleDecls.cpp:
```diff
@@ -483,6 +477,58 @@ static void
 }
 
+/// Trigger synthesizing implicit member declarations to make them "visible".
+static void synthesizeMemberDeclsForLookup(NominalTypeDecl *NTD,
```
How is this triggered during normal type-checking? This seems kind of brittle to live in lookup-visible-decls, although admittedly this whole file is basically like that...

rintaro replied:
If the type is declared in the primary files, they are synthesized in `TypeChecker::typeCheckDecl()`(`DeclChecker`) or in `typeCheckFunctionsAndExternalDecls()`.

But if it's in the non-primary files, AFAIU, they are lazily synthesized on demand in `DeclContext::lookupQualified()` 
https://github.com/apple/swift/blob/bd971507bae5a12a2b40365e795799037ec59fcc/lib/AST/NameLookup.cpp#L1573-L1579
https://github.com/apple/swift/blob/bd971507bae5a12a2b40365e795799037ec59fcc/lib/Sema/TypeChecker.h#L1018-L1024

I can move this function into `TypeChecker::`, but I don't think it's worth...

rintaro commented:
Resolved conflicts.
@swift-ci Please smoke test

slavapestov commented on lib/IDE/CodeCompletion.cpp:
```diff
@@ -4364,9 +4400,11 @@ class CompletionOverrideLookup : public swift::VisibleDeclConsumer {
+    if (D->isFinal())
+      return;
+
+    // A 'class' member with an initial value cannot be overriden either.
```
I'm not sure I understand. Can you give an example? Normally stored properties can be overridden with computed properties as long as they're not final.

rintaro replied:
I believe this meant something like:
```swift
class C {
  class var foo = 1
}
```
But that's compile error so ... not sure.
Here, I'm just changing `D->getAttrs().hasAttribute<HasInitialValueAttr>()` to `isa<VarDecl>(D) && cast<VarDecl>(D)->hasInitialValue()`

slavapestov replied:
My point was that the presence of an initial value expression does not have any bearing on whether the declaration can be overridden or not. The comment above is incorrect.

I think the only relevant check here is whether the property is final (`static` is equivalent to `class final`).

rintaro replied:
Right, I just removed this check.

slavapestov commented on lib/IDE/CodeCompletion.cpp:
```diff
@@ -4138,10 +4138,46 @@ class CompletionOverrideLookup : public swift::VisibleDeclConsumer {
+
+          auto Witness = Conformance->getWitnessDecl(Requirement);
+          if (Witness == VD)
+            Access = std::max(Access, Requirement->getFormalAccess());
```
The access of the requirement is the same as the access of the protocol, so I believe this entire loop and the search for the witness of 'VD' can go away.

rintaro replied:
You mean like this?
```cpp
    if (Access < AccessLevel::Public &&
        !isa<ProtocolDecl>(VD->getDeclContext())) {
      for (auto Conformance : CurrentNominal->getAllConformances()) {
        auto Proto = Conformance->getProtocol();
        Access = std::max(Access, Proto->getFormalAccess());
      }
    }
```
This may cause unnecessary `public` when the type conforms to two or more protocols. Like
```swift
public protocol P1 { }
protocol P2 {
  func bar()
}
public class B {
  func bar() {}
}
public class C: B, P1, P2 {
  <complete>
}
```
In this case when overriding `bar()`, `C.bar()` is not needed to be `public`.

slavapestov commented on lib/IDE/ExprContextAnalysis.cpp:
```diff
@@ -91,15 +91,35 @@ void typeCheckContextImpl(DeclContext *DC, SourceLoc Loc) {
+  // protocols in the AST.
+  (void) DC->getASTContext().getStdlibModule();
+
+  bindExtensions(*DC->getParentSourceFile());
```
Maybe the getStdlibModule() and bindExtensions() calls should be done as part of performParseAndNameBindingOnly()?

rintaro replied:
Made `bindExtensions()` a part of `performParseAndNameBindingOnly()`.

It turned out that `getStdlibModule()` is meaningless. Since `ASTContext::getStdlibModule(bool loadIfAbsent = false)` is called with default `loadIfAbsent = false`, it doesn't do anything. So I just removed it from `performTypechecking()`

slavapestov commented on lib/Sema/LookupVisibleDecls.cpp:
```diff
@@ -478,6 +474,56 @@ static void
+        Conformance->getRootConformance());
+    if (!NormalConformance)
+      continue;
+    for (auto Member : Proto->getMembers()) {
```
You can replace this loop with calls to `Conformance->forEachTypeWitness()` and `forEachValueWitness()`, passing in `useResolver=true`.

rintaro replied:
Done.

slavapestov commented on lib/Sema/LookupVisibleDecls.cpp:
```diff
@@ -478,6 +474,56 @@ static void
+    for (auto Member : IDC->getMembers()) {
+      if (auto var = dyn_cast<VarDecl>(Member)) {
+        if (var->hasAttachedPropertyWrapper()) {
+          auto sourceFile = var->getDeclContext()->getParentSourceFile();
```
You can check if the IDC itself is a member of a source file outside the loop.

rintaro replied:
Thanks, done!

slavapestov commented on lib/Sema/LookupVisibleDecls.cpp:
```diff
@@ -478,6 +474,56 @@ static void
+                                           const DeclContext *DC) {
+  // Synthesize the memberwise initializer for structs or default initializer
+  // for classes.
+  if (!NTD->hasInterfaceType())
```
It's possible that an interface type has been set, but implicit constructors were not added (and vice versa). You can unconditinally call `addImplicitConstructors()` here; it's idempotent if called multiple times.

rintaro replied:
IIRC, that causes circular dependency error. I will try that again.

rintaro replied:
Here's the cyclic dependency error when enabling typo correction. How can we fix this?

```
$ cat test.swift
struct Broken {
  var b = True
}

$ swift -frontend -typecheck -typo-correction-limit 10 -debug-cycles test.swift
test.swift:1:8: error: circular reference
struct Broken {
       ^
test.swift:2:3: note: through reference here
  var b = True
  ^
test.swift:2:7: note: through reference here
  var b = True
      ^
test.swift:2:7: note: through reference here
  var b = True
      ^
===CYCLE DETECTED===
 `--SynthesizeMemberwiseInitRequest(test.(file).Broken@test.swift:1:8)
     `--ParseMembersRequest(test.(file).Broken@test.swift:1:8) -> {(unknown decl), test.(file).Broken.b@test.swift:2:7}
     `--StorageImplInfoRequest(test.(file).Broken.b@test.swift:2:7)
     |   `--AttachedPropertyWrappersRequest(test.(file).Broken.b@test.swift:2:7)
     `--StorageImplInfoRequest(test.(file).Broken.b@test.swift:2:7)
     |   `--AttachedPropertyWrappersRequest(test.(file).Broken.b@test.swift:2:7)
     `--AccessLevelRequest(test.(file).Broken.b@test.swift:2:7)
     `--AccessLevelRequest(test.(file).Broken.b@test.swift:2:7)
     `--InterfaceTypeRequest(test.(file).Broken.b@test.swift:2:7)
         `--NamingPatternRequest(test.(file).Broken.b@test.swift:2:7)
             `--PatternBindingEntryRequest((unknown decl), 0)
                 `--AttachedPropertyWrappersRequest(test.(file).Broken.b@test.swift:2:7)
 ...
                 `--ParseMembersRequest(test.(file).Broken@test.swift:1:8) -> {(unknown decl), test.(file).Broken.b@test.swift:2:7}
                 `--GenericParamListRequest(test.(file).Broken@test.swift:1:8)
                 `--GenericSignatureRequest(test.(file).Broken@test.swift:1:8)
                 |   `--GenericParamListRequest(test.(file).Broken@test.swift:1:8)
                 `--GenericParamListRequest(test.(file).Broken@test.swift:1:8)
                 `--InterfaceTypeRequest(test.(file).Broken@test.swift:1:8)
                 |   `--GenericParamListRequest(test.(file).Broken@test.swift:1:8)
                 |   `--GenericParamListRequest(test.(file).Broken@test.swift:1:8)
                 `--HasMemberwiseInitRequest(test.(file).Broken@test.swift:1:8) -> 1
                 |   `--InterfaceTypeRequest(test.(file).Broken@test.swift:1:8)
                 |   |   `--GenericParamListRequest(test.(file).Broken@test.swift:1:8)
                 |   |   `--GenericParamListRequest(test.(file).Broken@test.swift:1:8)
                 |   `--HasUserDefinedDesignatedInitRequest(test.(file).Broken@test.swift:1:8) -> 0
                 |   |   `--ParseMembersRequest(test.(file).Broken@test.swift:1:8) -> {(unknown decl), test.(file).Broken.b@test.swift:2:7}
                 |   `--ParseMembersRequest(test.(file).Broken@test.swift:1:8) -> {(unknown decl), test.(file).Broken.b@test.swift:2:7}
                 |   `--StorageImplInfoRequest(test.(file).Broken.b@test.swift:2:7)
                 |   |   `--AttachedPropertyWrappersRequest(test.(file).Broken.b@test.swift:2:7)
                 |   `--StorageImplInfoRequest(test.(file).Broken.b@test.swift:2:7)
                 |   |   `--AttachedPropertyWrappersRequest(test.(file).Broken.b@test.swift:2:7)
                 |   `--AccessLevelRequest(test.(file).Broken.b@test.swift:2:7)
                 `--SynthesizeMemberwiseInitRequest(test.(file).Broken@test.swift:1:8) (cyclic dependency)
```

slavapestov replied:
I see. It might be better then to check to explicitly check if an SynthesizeMemberwiseInitRequest is active, instead of using `hasInterfaceType()`, ie,

```
VD->getASTContext().evaluator.hasActiveRequest(
            SynthesizeMemberwiseInitRequest{...})
```

slavapestov commented on lib/Sema/LookupVisibleDecls.cpp:
```diff
@@ -478,6 +474,56 @@ static void
+          resolver->resolveTypeWitness(NormalConformance, ATD);
+      } else if (auto *VD = dyn_cast<ValueDecl>(Member)) {
+        if (!NormalConformance->hasWitness(VD))
+          resolver->resolveWitness(NormalConformance, VD);
```
Can we do this all this inside lookupDeclsFromProtocolsBeingConformedTo() instead?

rintaro replied:
I don't think so because this has to be done before `lookupTypeMembers()` which is called before `lookupDeclsFromProtocolsBeingConformedTo()`.

slavapestov commented on lib/Sema/LookupVisibleDecls.cpp:
```diff
@@ -478,6 +474,56 @@ static void
+  };
+  synthesizePropertyWrappers(NTD);
+  for (auto ED : NTD->getExtensions())
+    synthesizePropertyWrappers(ED);
```
I might be cleaner to visit the extension's members in doGlobalExtensionLookup()

rintaro replied:
Done

rintaro committed:
Factor out synthesizePropertyWrapperStorageWrapperProperties()

rintaro committed:
Utilize forEachTypeWitness()/forEachValueWitness()

rintaro committed:
Do 'bindExtensions()' as a part of 'performParseAndNameBindingOnly()'

rintaro commented:
@swift-ci Please smoke test

DougGregor reviewed (commented):
I don't have much to add to Slava and Ben's feedback, but I love the form of this. It's great to see us taking advantage of more laziness in the type checker.

slavapestov commented on lib/IDE/CodeCompletion.cpp:
```diff
@@ -4138,10 +4138,41 @@ class CompletionOverrideLookup : public swift::VisibleDeclConsumer {
+        Conformance->getRootConformance()->forEachValueWitness(
+            [&](ValueDecl *req, Witness witness) {
+              if (witness.getDecl() == VD)
+                Access = std::max(Access, req->getFormalAccess());
```
I still think it's better to write `Conformance->getProtocol()->getFormalAccess()` here.

rintaro committed:
Workaround cyclic dependency issue in memberwise initializer synthesis

rintaro committed:
[CodeCompletion] Remove unsound check in override completion

In override completion, we didn't use to emit 'class var' with the initial
expression. However, having initial expression does not have any bearing
on whether the declaration is overridable or not.

rintaro commented:
@swift-ci Please smoke test

rintaro commented:
@swift-ci Please smoke test
